src/datasets/user-wgn.dataset.ts

```diff
@@ -0,0 +1,176 @@
+}
+
+export const UserWGNDatasetDefaultConfig: UserWGNDatasetConfig = {
+  avgHits: 10000,
+  hitStdDev: 100,
+  avgUsers: 100,
+  userStdDev: 1,
+  avgSessionsPerUser: 5,
+  sessionsPerUserStdDev: 3,
+};
+
+// TODO: Refactoring
+export function UserWGNDataset(config: UserWGNDatasetConfig): Dataset {
+  const categories = [countryCategory, browserCategory, sourceCategory];
+  const measures = [activeUserMeasure, revenueMeasure, eventCountMeasure];
+
+  // TODO: configurable startDate and endDate for data generation in DatasetPreference
+  const endDate = new Date();
+  const startDate = new Date(endDate.getTime() - 30 * DAY);
+
+  const dataCube = generateCube(categories, measures, {
+    ...UserWGNDatasetDefaultConfig,
+    ...config,
+  });
+
+  const measureNames = measures.map(measure => measure.name);
+
+  const pureMeasureNames = measureNames.filter(measureName => !isCompoundMeasure(measureName));
+  const compoundMeasures = measureNames.filter(isCompoundMeasure).map(destructureCompoundMeasure);
+  const windowSizes = compoundMeasures
+    .filter((measure: CompoundMeasure): measure is RollingMeasure => measure.type === CompoundMeasureType.ROLLING)
+    .map(({ windowSize }) => windowSize);
+  const periodOffsets = compoundMeasures
+    .filter((measure: CompoundMeasure): measure is PeriodOverPeriodMeasure => measure.type === CompoundMeasureType.PERIOD_OVER_PERIOD)
+    .map(({ periodOffset }) => periodOffset);
+
+  const dateCategoryName = 'date';
+  const duration = endDate.getTime() - startDate.getTime();
+  const maxWindowSize = Math.max(0, ...windowSizes);
+  const dateRanges: [Date, Date][] = [0, ...periodOffsets].map(periodOffset => {
+    const periodStart = startDate.getTime() + periodOffset;
+    const rangeStartDate = new Date(periodStart - maxWindowSize);
+    const rangeEndDate = new Date(periodStart + duration);
+    return [rangeStartDate, rangeEndDate];
+  });
+  const dateFilter = inOneOfDateRanges(dateRanges, { excludeStart: true });
+
+  const rawData = dataCube.getDataFor({
+    categoryNames: [dateCategoryName, ...pureMeasureNames],
+    measureNames,
+    filters: [dateFilter],
+    sortBy: [dateCategoryName],
+  });
+
+  const data = rawData.filter(datum => {
+    const { date } = datum.categories;
+    return startDate < date && date <= endDate;
+  });
+
+  // TODO: Refactoring
+  for (const measureName of pureMeasureNames) {
+    for (const rollingUnit of windowSizes) {
+      attachRollingMeasure(rawData, data, measureName, rollingUnit);
+    }
+    for (const periodOffset of periodOffsets) {
+      attachPeriodOverPeriodMeasure(rawData, data, measureName, periodOffset);
+    }
+  }
+
+  const makeLineChartMeta = (measureName: string): LineChartMeta => {
+    const rows = data.map(row => ({
+      x: row.categories.date,
+      y: row.values[measureName],
+    }));
+
+    const query = (options: LineChartQueryOptions) => ([{
+      points: rows.filter(xBetweenDates(...options.range)),
+    }]);

```

Comment:
I think the data cube querying behavior (line 59 - 107) should be happening within this `query` method.

When we think of `DataCube` as the backend server of GA, it's like fetching the entire data without knowing which portion of the data is actually needed.

Reply:
It is for performance concerns. As `queryOptions` has no effect on `rows` (line 103), we can prepare them before the `query` function is called and reuse them when applying different `queryOptions`. We just need to "mock" the query service.

By the way you made me notice that I should move line 103 outside function `createLineChartMeta`.

Reply:
Let's say the range options are 1 month and 1 year long. Are we going to query in advance for the entire year data when the initial range option is set to be only 1 month long?

Reply:
GA similarly makes a new API call (or "query") to the server (or "data cube") every time the range changes.

Reply:
Considering users don't change range options as often, initially querying small amount of data would rather enhance performance.

Reply:
For this dataset: yes. Since these implementations have similar memory consumption and are both correct, which means the `query` function produces exact same output, I would choose the one with less execution time. Moreover, other datasets can have their own `query` implementation.

Reply:
Or we can have some utilities to create query function with lazy query and caches.

Reply:
Both do behave correctly, but the current implementation would consume way more memory.

To roughly estimate, the current implementation would have `rawData` with the length of 365, and my suggested implementation would have the length of 30.

Even if they were to have similar memory consumption, I would mimic how GA is handling it, considering the goal of this project.

Reply:
> Or we can have some utilities to create query function with lazy query and caches later. What do you think?

Yeah I agree that we should make it a util since many of datasets/metas would have a similar query function.

Reply:
> Both do behave correctly, but the current implementation would consume way more memory.
> 
> To roughly estimate, the current implementation would have `rawData` with the length of 365, and my suggested implementation would have the length of 30.
> 
> Even if they were to have similar memory consumption, I would mimic how GA is handling it, considering the goal of this project.

The memory consumption should be dominated by object `dataCube`. That's why I said similar memory consumption.
 
The real-world GA data can be used in this project by creating a dataset that sends requests to the backend service in query functions. It has nothing to do with the fake dataset generated by data generator.

Reply:
It is true that the memory consumption is dominated by `dataCube`, but by querying the entire `dataCube`, the current implementation doubles the memory consumption.

Reply:
> > Or we can have some utilities to create query function with lazy query and caches later. What do you think?
> 
> Yeah I agree that we should make it a util since many of datasets/metas would have a similar query function.

How about working on that in a separate PR? After you merge it to #45 and add those compounded measures.
You can modify the implementation when merging it, and I also hope you to refactor the `create` function consider the change of the entire design.

Reply:
Okay, I'll refactor the `create` function in #45. Are you planning to make further updates to `user-white-noise.dataset.ts` in this PR?

Reply:
Actually I think it's better to refactor the function in a separate branch (#X), since this refactoring would be far from what #45 is focusing on. I'll merge #X into this PR under your review, and then merge this PR into #45.  How do you think?

Reply:
Great. I have added cards to the project board.